p2p.go

```diff
@@ -506,28 +517,40 @@ func (h *Host) Connect(ctx context.Context, target peerstore.PeerInfo) error {
 }
 
 // Broadcast sends a message to the hosts who subscribe the topic
-func (h *Host) Broadcast(topic string, data []byte) error {
+func (h *Host) Broadcast(ctx context.Context, topic string, data []byte) error {
 	pub, ok := h.pubs[topic]
 	if !ok {
 		return nil
 	}
-	return pub.Publish(topic, data)
+	return pub.Publish(ctx, data)
 }
 
 // Unicast sends a message to a peer on the given address
-func (h *Host) Unicast(ctx context.Context, target peerstore.PeerInfo, topic string, data []byte) error {
-	if err := h.Connect(ctx, target); err != nil {
+func (h *Host) Unicast(ctx context.Context, target core.PeerAddrInfo, topic string, data []byte) error {
+	if h.unicastBlocklist.Blocked(target.ID, time.Now()) {
+		return errors.New("peer is in blocklist at this moment")
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			h.unicastBlocklist.Add(target.ID, time.Now())
+		}
+	}()
+	if err = h.Connect(ctx, target); err != nil {
 		return err
 	}
 	stream, err := h.host.NewStream(ctx, target.ID, protocol.ID(topic))
 	if err != nil {
 		return err
 	}
-	defer func() { err = stream.Close() }()
-	if _, err = stream.Write(data); err != nil {
-		return err
-	}
-	return nil
+	defer func() {
+		if err := stream.Close(); err != nil {
+			Logger().Error("Error when closing a unicast stream.", zap.Error(err))
+		}
+	}()
+	_, err = stream.Write(data)
```

Comment:
the two nested "defer" functions both with "err" is so hard to understand. We may have a help function "unicast" to contain the unicast logic, including "connect-> new stream->write->close", and return an error, which will be used to determine whether the target should be added into blocklist

Reply:
updated